### PR TITLE
Flow Trigger Bundle

### PR DESCRIPTION
**Pull request by programmarchy**
Implemented panel and interface.

However, blocked on fully implementing the display. My interpretation of the spec was that the display should work similar to the "User" display that displays an avatar and/or name for a m2o relation, but pointed to `directus_flows` rather than `directus_users`. However, this part of the spec I didn't find a way to implement:

> pass in the ID of the item on which it has been triggered as the sole item in the keys array

The display context has access to the related _flow_ via `props.value`, however, I wasn't able to find any way to access the ID of the item. The flows sidebar uses `props.selection` to make this relationship, but there's no equivalent for a display extension. Is this possible, or am I missing something? Please advise.

<details>

<summary>Panel screenshots</summary>

panel-1

panel-2

panel-3

panel-4

</details>

<details>

<summary>Interface screenshots</summary>

interface-1

interface-2

interface-3

</details>

<details>

<summary>Display screenshots</summary>

display-1

display-2

</details>

### CODE REVIEW

File: packages/flow-trigger-bundle/src/flow-trigger-display/display.vue [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+	if (!flow) {
+		return;
+	}
+
+	const flowId = flow.id;
+
+	runningFlows.value = [...runningFlows.value, flowId];
+
+	const collection = props.collection;
+	const keys = [];
```
Comment (programmarchy) [inline packages/flow-trigger-bundle/src/flow-trigger-display/display.vue:128]: How do I access the primary key of the item related to this display?

Comment (ComfortablyCoding) [reply packages/flow-trigger-bundle/src/flow-trigger-display/display.vue:128]: Unfortunately not possible for displays at this time

Comment (ComfortablyCoding) [reply packages/flow-trigger-bundle/src/flow-trigger-display/display.vue:128]: Per @phazonoverload the display can be removed from the PR for now.

Comment (phazonoverload) [reply packages/flow-trigger-bundle/src/flow-trigger-display/display.vue:128]: I thought the displays got passed the ID when originally writing the spec. I was wrong. Let's remove it for now, budget unchanged, and then we will swing back to it at another point in time as a new piece of work

Comment (programmarchy) [reply packages/flow-trigger-bundle/src/flow-trigger-display/display.vue]: @ComfortablyCoding @phazonoverload Sounds good. Removed the display.

**Review by ComfortablyCoding — CHANGES_REQUESTED** [review]
Overall very well done!

### Feedback
1. I see duplicated functions/code (e.g.`getFlow`) between files as this is a bundle we should have shared functions/logic where possible.
2. For the panel it seems I cannot add a flow where I have read permission to add least one of the collections that are triggered. I also cannot click the button to trigger the flow.
3. For cases where the button is not clickable (e.g. I do NOT have read access to at least one collection) it should be set as disabled if possible.
4. A readme is missing

**Comment by programmarchy** [comment]
> Overall very well done!

Thanks!

> ### Feedback
> 1. I see duplicated functions/code (e.g.`getFlow`) between files as this is a bundle we should have shared functions/logic where possible.

Agreed 👍 

> 2. For the panel it seems I cannot add a flow where I have read permission to add least one of the collections that are triggered. I also cannot click the button to trigger the flow.

One caveat you might be running into is that the collection must also be selected in the flow settings. I found that if you try to run a flow on a collection that is not selected in the flow settings, then you'll get a forbidden error, even if you have access to that collection. Therefore, during configuration, the collection dropdown is filtered down to collections selected in the flow settings.

If that's not the case, could you share your setup with me? Maybe attach the data.db?

> 3. For cases where the button is not clickable (e.g. I do NOT have read access to at least one collection) it should be set as disabled if possible.

In those cases, the button shouldn't be visible currently. I only added that in the latest commit, though.

Again, if you're seeing something differently, please share your setup so I can reproduce.

> 4. A readme is missing

👍  I'll write something up.

File: packages/flow-trigger-bundle/src/system-flow-trigger-form/interface.vue [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+							choices: collections.value,
+							allowNone: false,
+						},
+					},
+				},
+				{
+					field: 'keys',
+					name: 'IDs',
+					type: 'csv',
+					meta: {
```
[the remark this answers is not in the document]
Comment (programmarchy) [reply packages/flow-trigger-bundle/src/system-flow-trigger-form/interface.vue:118]: The only reason I didn't do that was to match the language used in the Read/Update/Delete flow operations. No reason not to make the improvement in this extension though. Done!

**Comment by programmarchy** [comment]
@ComfortablyCoding Couple questions for you in the above comment about 2 and 3. In the meantime I'll work on 1 and 4.

**Comment by ComfortablyCoding** [comment]
> @ComfortablyCoding Couple questions for you in the above comment about 2 and 3. In the meantime I'll work on 1 and 4.

@programmarchy see demo below, open to doing screen share displaying the issue if needed.

### Number 2
https://github.com/user-attachments/assets/58ad9c92-046a-49fe-9e51-af53cdd8a8df

### Number 3
https://github.com/user-attachments/assets/98c03705-1ca2-4bd0-918d-194f4d4b35f5

### User Permissions
Screenshot 2024-08-27 at 3 28 54 PM

### Flow Settings
Screenshot 2024-08-27 at 3 36 35 PM

**Comment by ComfortablyCoding** [comment]
An additional point of feedback: I am only able to select one collection for the flow button, per the reference multiple collection selection should be supported

**Comment by programmarchy** [comment]
@ComfortablyCoding Sorry for the delay. Missed the GitHub notification.

Could you please screenshot one more item? Please open the panel showing "No manual flows are available" and screenshot the raw value via the "Flow" label dropdown. My guess is that it's set to something other than "article".

Screenshot 2024-08-30 at 12 01 44 PM

Screenshot 2024-08-30 at 12 07 11 PM

**Comment by programmarchy** [comment]
@ComfortablyCoding

> An additional point of feedback: I am only able to select one collection for the flow button, per the reference multiple collection selection should be supported

Not sure I understand. Here's the part of the spec I was interpreting:

> It’s anticipated that users must either provide a hardcoded item ID to be passed in the keys array, or use the dynamic value from a relational global variable panel in the same dashboard.

Given the requirement that users need to provide keys at configuration time (hardcoded or `{{global_var}}`), those keys must map to a specific collection. If you could select multiple collections in the configuration step, how would the button know which collection to use at when clicking the button on the panel?

**Comment by ComfortablyCoding** [comment]
> @ComfortablyCoding Sorry for the delay. Missed the GitHub notification.
> 
> Could you please screenshot one more item? Please open the panel showing "No manual flows are available" and screenshot the raw value via the "Flow" label dropdown. My guess is that it's set to something other than "article".
> 

I discovered the issue is that the user did not have `read` for `directus_flows` enabled. That seems to have resolved this issue, we should probably mention that in the issue.

Screenshot 2024-08-30 at 2 03 32 PM

**Comment by programmarchy** [comment]
@ComfortablyCoding Added a readme and did some refactoring to share common logic in a composable. I also moved the "system" interface into a separate folder. Not sure if you'll need to do any finagling to make Directus see the new extension path.

**Comment by programmarchy** [comment]
> I discovered the issue is that the user did not have read for directus_flows enabled.

@ComfortablyCoding Better yet, I could detect if the user doesn't have permission and display a better notice in this case.

**Comment by ComfortablyCoding** [comment]
> > I discovered the issue is that the user did not have read for directus_flows enabled. That seems to have resolved this issue, we should probably mention this requirement in the readme/instructions.
> 
> @ComfortablyCoding Better yet, I could detect if the user doesn't have permission and display a better notice in this case.

That would be great!

**Comment by ComfortablyCoding** [comment]
> @ComfortablyCoding Added a readme and did some refactoring to share common logic in a composable.

👍 

> I also moved the "system" interface into a separate folder. Not sure if you'll need to do any finagling to make Directus see the new extension path.

hmm, it should be fine as path is picked up by whatever is defined in `package.json`. I will double check

**Comment by phazonoverload** [comment]
> Given the requirement that users need to provide keys at configuration time (hardcoded or {{global_var}}), those keys must map to a specific collection. If you could select multiple collections in the configuration step, how would the button know which collection to use at when clicking the button on the panel?

I think there are approached to achieve this, but I'm happy with the current approach for the time being. Perhaps an enhancement if it's later requested.

**Comment by programmarchy** [comment]
> I think there are approached to achieve this, but I'm happy with the current approach for the time being. Perhaps an enhancement if it's later requested.

Sounds good. One thing we could do is leave collection + keys optional at configuration, and if they are empty at run time, then prompt the user to pick collection + keys using a `collection-item-dropdown`.

**Comment by programmarchy** [comment]
@ComfortablyCoding Added the notice to cover the case where a user doesn't have permission to read flows so this PR is ready for re-review.

**Review by ComfortablyCoding — DISMISSED** [review]
LGTM 🚀 , thanks @programmarchy !

**Review by phazonoverload — APPROVED** [review]

**Comment by Yuggs** [comment]
Hi,
I've been struggling to build an extension like this for quite some time now, thank you guys for your work!
I'm not sure this is really an issue or just that this is not the way the interface is supposed to work, but when I trigger a flow using the interface on an item page, it doesn't seem to refresh the page/fields that were modified.
Is it something that would be possible?

**Pull request merged**